ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/NoSameLineStatementAsOfBlockStartOrEnd.kt

```diff
@@ -0,0 +1,90 @@
+package com.pinterest.ktlint.ruleset.standard.rules
+
+import com.pinterest.ktlint.rule.engine.core.api.ElementType
+import com.pinterest.ktlint.rule.engine.core.api.IndentConfig
+import com.pinterest.ktlint.rule.engine.core.api.editorconfig.EditorConfig
+import com.pinterest.ktlint.rule.engine.core.api.editorconfig.INDENT_SIZE_PROPERTY
+import com.pinterest.ktlint.rule.engine.core.api.editorconfig.INDENT_STYLE_PROPERTY
+import com.pinterest.ktlint.rule.engine.core.api.indent
+import com.pinterest.ktlint.rule.engine.core.api.nextCodeLeaf
+import com.pinterest.ktlint.rule.engine.core.api.noNewLineInClosedRange
+import com.pinterest.ktlint.rule.engine.core.api.prevCodeLeaf
+import com.pinterest.ktlint.rule.engine.core.api.upsertWhitespaceAfterMe
+import com.pinterest.ktlint.rule.engine.core.api.upsertWhitespaceBeforeMe
+import com.pinterest.ktlint.ruleset.standard.StandardRule
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+
+public class NoSameLineStatementAsOfBlockStartOrEnd
+    : StandardRule(
+    "no-same-line-statement-as-of-block-start-or-end",
+    usesEditorConfigProperties = setOf(
+        INDENT_SIZE_PROPERTY,
+        INDENT_STYLE_PROPERTY,
+    )
+) {
+    private var indentConfig = IndentConfig.DEFAULT_INDENT_CONFIG
+
+    override fun beforeFirstNode(editorConfig: EditorConfig) {
+        indentConfig =
+            IndentConfig(
+                indentStyle = editorConfig[INDENT_STYLE_PROPERTY],
+                tabWidth = editorConfig[INDENT_SIZE_PROPERTY],
+            )
+    }
+
+    override fun beforeVisitChildNodes(
+        node: ASTNode,
+        autoCorrect: Boolean,
+        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit,
+    ) {
+        if (node.elementType == ElementType.FUN) {
```

Comment:
Do not restrict it to function bodies. I think, it is safe to apply this on any block. But it might be needed to add an exception to this, which would allow a block starting and ending on the same line. For example in definition of a function literal:
```
val foo = { bar() }
```

Reply:
What about
```
fun foo() { println() }
```
should we allow it? Currently, I have disallowed it. But if we are allowing above we should also allow single line functions with block

Reply:
It should be disallowed. The example `val foo = { bar() }` is a function literal and will not have element type 'BLOCK'. The IntelliJ Plugin `PsiViewer` is a very helpful too to check how the AST of a code sample is constructed.

Reply:
Got it. Basically, we will allow single-line function literal but will NOT allow single-line function body. Thanks for the clarification.